### PR TITLE
Feature/xref optimisation

### PR DESCRIPTION
**Pull request by magaliruffier**
## Requirements

- Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion;
- Review the [contributing guidelines](https://github.com/Ensembl/ensembl/blob/master/CONTRIBUTING.md#why-could-my-pull-request-be-rejected) for this repository; remember in particular:
    - do not modify code without testing for regression
    - provide simple unit tests to test the changes
    - if you change the schema you must patch the test databases as well, see [Updating the schema](https://github.com/Ensembl/ensembl/blob/master/CONTRIBUTING.md#updating-the-schema)
    - the PR must not fail unit testing

## Description

The proposed changes simplify the stage where data is copied from the xref database into the core database.

## Use case

Some of the queries run are quite heavy on the MySQL server. When multiple species are being run in parallel, these queries can get stuck on a heavily loaded server.
As they deal with data that is not used in the core database, removing these steps from the pipeline makes the code simpler and less onerous to run.

## Benefits

Less contention on the server for multiple running jobs.

## Possible Drawbacks

Historical unmapped entries are not conserved.

## Testing

_Have you added/modified unit tests to test the changes?_
No test cases available but the pipeline was run on about 20 species with and without the proposed changes.
The results between runs are comparable.
The mapping stage runs faster with the changes

_If so, do the tests pass/fail?_

_Have you run the entire test suite and no regression was detected?_

### CODE REVIEW

File: misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm [hunk 1]
```diff
@@ -65,21 +65,18 @@ sub update{
...
-DELETE ontology_xref, object_xref, xref, dependent_xref
-FROM ontology_xref, object_xref, xref 
-LEFT JOIN dependent_xref on xref_id = dependent_xref_id
-WHERE ontology_xref.object_xref_id = object_xref.object_xref_id AND object_xref.xref_id = xref.xref_id
-SQL
+  $sql = "DELETE ox FROM object_xref ox, xref x, external_db e WHERE x.xref_id = ox.xref_id AND x.external_db_id =e.external_db_id and e.db_name = 'GO'";
   $sth = $core_dbi->prepare($sql);
-  $affected_rows = $sth->execute();
-  print "\tDeleted $affected_rows PROJECTED ontology_xref row(s)\n" if $verbose;
+  $sth->execute();
```
Comment (ens-bwalts) [inline misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm:79]: Might be nice to confirm this deletion in a print statement, either by printing here or summing up the rows with the delete at lines 81-83

**Review by ens-bwalts — APPROVED** [review]
Otherwise this looks OK

File: misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm [hunk 2]
```diff
@@ -65,21 +65,18 @@ sub update{
...
   print "\tDeleted $affected_rows PROJECTED external_synonym row(s)\n" if $verbose;
-
-  # Delete all ontologies, as they are done by a separate pipeline
-  $sql = <<SQL;
-DELETE ontology_xref, object_xref, xref, dependent_xref
-FROM ontology_xref, object_xref, xref 
-LEFT JOIN dependent_xref on xref_id = dependent_xref_id
-WHERE ontology_xref.object_xref_id = object_xref.object_xref_id AND object_xref.xref_id = xref.xref_id
-SQL
+  $sql = "DELETE ox FROM object_xref ox, xref x, external_db e WHERE x.xref_id = ox.xref_id AND x.external_db_id =e.external_db_id and e.db_name = 'GO'";
```
Comment (ameya1981) [inline misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm:77]: Is the intention to delete only from object_xref or from other tables too?
In the earlier version, alongwith ontology_xref, object_xref, xref and dependent_xref records were deleted too(?) for all ontologies.
Not sure if the other deletions(if still required) are handled in some other fashion.

File: misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm [hunk 3]
```diff
@@ -65,21 +65,18 @@ sub update{
   #####################################
   print "Deleting all PROJECTIONs from this database\n" if $verbose;
 
-  my $sql = "DELETE es FROM xref x, external_synonym es WHERE x.xref_id = es.xref_id and x.info_type = 'PROJECTION'";
+  my $sql = "TRUNCATE table ontology_xref";
```
Comment (ameya1981) [inline misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm:68]: Truncate would reset the auto_increment counter - hope that value is not being used anywhere else

File: misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm (hunk 1 above)
Comment (magaliruffier) [reply misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm:79]: as all the jobs are now run in the eHive pipeline, we don't actually see any of the delete statements, apart for the rare cases where a job is run in standalone mode, which is why it seemed unnecessary here
I believe a truncate job also does not actually say how many rows are affected, unlike a 'delete from table' query, which is why it is so much faster

File: misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm (hunk 2 above)
Comment (magaliruffier) [reply misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm:77]: the GPAD loading pipeline, which now deals with updating the ontology_xref entries, has a clean up step at the start and end of the pipeline
the step here is redundant with the GPAD loading pipeline and quite heavy on the database

File: misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm (hunk 3 above)
Comment (magaliruffier) [reply misc-scripts/xref_mapping/XrefMapper/XrefLoader.pm:68]: not to my knowledge

**Review by ameya1981 — APPROVED** [review]

**Pull request merged**